### ISSUE TITLE
Victorialogs: too big number of columns detected in the block

### ISSUE THREAD

**Issue by tripitakav** [VictoriaMetrics/VictoriaMetrics#4762]
### Describe the bug

panic: BUG: too big number of columns detected in the block: 10633; the number of columns mustn't exceed 10000

### To Reproduce

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app: vmlog
  name: vmlog
  namespace: ops
spec:
  serviceName: vmlog
  selector:
    matchLabels:
      app: vmlog
  replicas: 1
  template:
    metadata:
      labels:
        app: vmlog
    spec:
      containers:
        - name: vmlog
          image: "xxxxxxxxxxx/vmlog:hk-v0.3.1"
          imagePullPolicy: "IfNotPresent"
          args:
            - "--storageDataPath=/storage"
            - --retentionPeriod=8w
          ports:
            - name: http
              containerPort: 9428
          resources:
            limits:
              cpu: "2"
              memory: 10000Mi
            requests:
              cpu: 500m
              memory: 3000Mi
          volumeMounts:
            - name: vmlog-volume
              mountPath: /storage
      terminationGracePeriodSeconds: 60
  updateStrategy:
    type: RollingUpdate
  volumeClaimTemplates:
    - metadata:
        labels:
          app: vmlog-volume
        name: vmlog-volume
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 100Gi
        storageClassName: gp2
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: vmlog
  name: vmlog
  namespace: ops
spec:
  clusterIP: None
  ports:
    - port: 9428
      targetPort: http
      protocol: TCP
      name: http
  selector:
    app: vmlog
---
```

### Version

victoria-logs-20230721-040232-tags-v0.3.0-victorialogs-0-gda0caa1d34

### Logs
```
2023-08-02T03:06:52.709Z	info	VictoriaMetrics/lib/logger/flag.go:12	build version: victoria-logs-20230721-040232-tags-v0.3.0-victorialogs-0-gda0caa1d34
2023-08-02T03:06:52.709Z	info	VictoriaMetrics/lib/logger/flag.go:13	command-line flags
2023-08-02T03:06:52.709Z	info	VictoriaMetrics/lib/logger/flag.go:20	  -retentionPeriod="8w"
2023-08-02T03:06:52.709Z	info	VictoriaMetrics/lib/logger/flag.go:20	  -storageDataPath="/storage"
2023-08-02T03:06:52.709Z	info	VictoriaMetrics/app/victoria-logs/main.go:42	starting VictoriaLogs at ":9428"...
2023-08-02T03:06:52.709Z	info	VictoriaMetrics/app/vlstorage/main.go:52	opening storage at -storageDataPath=/storage
2023-08-02T03:06:52.709Z	info	VictoriaMetrics/lib/memory/memory.go:42	limiting caches to 6291456000 bytes, leaving 4194304000 bytes to the OS according to -memory.allowedPercent=60
2023-08-02T03:06:52.846Z	info	VictoriaMetrics/app/vlstorage/main.go:58	successfully opened storage in 0.136 seconds; partsCount: 64; blocksCount: 173263; rowsCount: 385566562; sizeBytes: 18594213336
2023-08-02T03:06:52.846Z	info	VictoriaMetrics/app/victoria-logs/main.go:50	started VictoriaLogs in 0.137 seconds; see https://docs.victoriametrics.com/VictoriaLogs/
2023-08-02T03:06:52.846Z	info	VictoriaMetrics/lib/httpserver/httpserver.go:96	starting http server at http://127.0.0.1:9428/
2023-08-02T03:06:52.846Z	info	VictoriaMetrics/lib/httpserver/httpserver.go:97	pprof handlers are exposed at http://127.0.0.1:9428/debug/pprof/
2023-08-02T03:06:54.691Z	warn	VictoriaMetrics/app/vlinsert/elasticsearch/elasticsearch.go:96	cannot decode log message #16 in /_bulk request: cannot parse timestamp: cannot parse timestamp in milliseconds from "1690945593819.000": strconv.ParseInt: parsing "1690945593819.000": invalid syntax
2023-08-02T03:06:54.754Z	warn	VictoriaMetrics/app/vlinsert/elasticsearch/elasticsearch.go:96	cannot decode log message #12 in /_bulk request: cannot parse timestamp: cannot parse timestamp in milliseconds from "1690945610015.000": strconv.ParseInt: parsing "1690945610015.000": invalid syntax
2023-08-02T03:06:55.006Z	warn	VictoriaMetrics/app/vlinsert/elasticsearch/elasticsearch.go:96	cannot decode log message #12 in /_bulk request: cannot parse timestamp: cannot parse timestamp in milliseconds from "1690945562038.000": strconv.ParseInt: parsing "1690945562038.000": invalid syntax
2023-08-02T03:06:55.104Z	warn	VictoriaMetrics/app/vlinsert/elasticsearch/elasticsearch.go:96	cannot decode log message #12 in /_bulk request: cannot parse timestamp: cannot parse timestamp in milliseconds from "1690945612946.787": strconv.ParseInt: parsing "1690945612946.787": invalid syntax
2023-08-02T03:06:55.904Z	warn	VictoriaMetrics/app/vlinsert/elasticsearch/elasticsearch.go:96	cannot decode log message #256 in /_bulk request: cannot parse timestamp: cannot parse timestamp in milliseconds from "1690945472.349": strconv.ParseInt: parsing "1690945472.349": invalid syntax
2023-08-02T03:06:55.941Z	warn	VictoriaMetrics/app/vlinsert/elasticsearch/elasticsearch.go:96	cannot decode log message #45 in /_bulk request: cannot parse timestamp: cannot parse timestamp in milliseconds from "1690945613973.076": strconv.ParseInt: parsing "1690945613973.076": invalid syntax
2023-08-02T03:06:55.943Z	panic	VictoriaMetrics/lib/logstorage/block.go:374	BUG: too big number of columns detected in the block: 10633; the number of columns mustn't exceed 10000
panic: BUG: too big number of columns detected in the block: 10633; the number of columns mustn't exceed 10000

goroutine 86 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logMessage({0xc44169, 0x5}, {0xc00009d500, 0x67}, 0x2?)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:281 +0xa96
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevelSkipframes(0x1, {0xc44169, 0x5}, {0xc7907e?, 0xb869a0?}, {0xc0006c9a40?, 0xc010916000?, 0xc010955460?})
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:138 +0x1a5
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevel(...)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:130
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.Panicf(...)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:126
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*block).sortColumnsByName(0xc0001128c0)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block.go:374 +0x98
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*block).MustInitFromRows(0xc0001128c0, {0xc000844000?, 0x39a, 0x2400}, {0xc000880000, 0x39a, 0x2400})
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block.go:232 +0x125
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamWriter).MustWriteRows(0xc0005c19e0?, 0xc00dba2000?, {0xc000844000, 0x39a, 0x2400}, {0xc000880000, 0x39a, 0x2400})
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_writer.go:228 +0xcb
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamMerger).mustFlushRows(0xc001fe8000)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_merger.go:236 +0x85
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamMerger).mustMergeRows(0xc001fe8000, 0xc00077eb40?)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_merger.go:214 +0x3f7
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamMerger).mustWriteBlock(0xc001fe8000, 0xc00077eb40, 0xc0001c0310?)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_merger.go:146 +0x8c
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.mustMergeBlockStreams(0xc000616340?, 0xc00046b840?, {0xc0001c0310?, 0xe, 0xe}, 0x7f6eaf542118?)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_merger.go:23 +0xec
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).mustMergeParts(0xc0000fbad0, {0xc0001c0150, 0xe, 0xe}, 0x0)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:391 +0x4c5
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).mustMergeExistingParts(0xc0000fbad0)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:287 +0x1ca
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).startMergeWorkerLocked.func1()
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:238 +0x38
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).startMergeWorkerLocked
	github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:236 +0xa5
```

**Comment by zekker6** [comment]
@tripitakav Could you please share configuration which you are using to ship logs to VictoriaLogs?

```
2023-08-02T03:06:55.006Z warn VictoriaMetrics/app/vlinsert/elasticsearch/elasticsearch.go:96 cannot decode log message https://github.com/VictoriaMetrics/VictoriaMetrics/issues/12 in /_bulk request: cannot parse timestamp: cannot parse timestamp in milliseconds from "1690945562038.000": strconv.ParseInt: parsing "1690945562038.000": invalid syntax
```
This error message suggests that format of date is not expected. Looks like it has trailing zeroes representing nanoseconds, could you confirm if this format is parsed by Elastic?
If so, we will also add support for this format.

```
 2023-08-02T03:06:55.943Z panic VictoriaMetrics/lib/logstorage/block.go:374 BUG: too big number of columns detected in the block: 10633; the number of columns mustn't exceed 10000
panic: BUG: too big number of columns detected in the block: 10633; the number of columns mustn't exceed 10000
```
This error means that number of a single log item contained more than 10000 fields, is this expected?
Most likely this is caused by some misconfiguration on the log shipper side.
In order to avoid such issue we're going to add a limit on number of unique log fields on ingestion, so that such log lines will be skipped by default.

**Comment by tripitakav** [comment]
@zekker6 Here is my configuration, we use vector to send log to VictoriaLogs
`yaml
[sinks.vlogs]
  inputs = ["droplogs"]
  type = "elasticsearch"
  endpoints = [ "http://vmlog.ops.svc:9428/insert/elasticsearch/" ]
  mode = "bulk"
  api_version = "v8"
  healthcheck.enabled = false
  [sinks.vlogs.query]
    _msg_field = "message"
    _time_field = "timestamp"
    _stream_fields = "app,pod_name,container_name"
`


I don't know if Elastic can parse it. We are using Loki, it can parse it rightly!

**Comment by zekker6** [comment]
@tripitakav Could you check if using Loki ingestion protocol for VictoriaLogs does not produce those errors for you?

Here is a configuration example:
```
[sinks.vlogs]
  type = "loki"
  inputs = [ "droplogs" ]
  endpoint = "http://vmlog.ops.svc:9428/insert/loki/"
  compression = "gzip"
  path = "/api/v1/push?_stream_fields=app,pod_name,container_name&_msg_field=message&_time_field=timestamp"
  tenant_id = "0:0"
  encoding.codec = "raw_message"
```

**Comment by tripitakav** [comment]
@zekker6 Thanks! I using Loki ingestion protocol for VictoriaLogs don't produce those errors. But there are some warning logs，It should be that my sinks _[sinks.vlogs]_ did not disable healthscheck
![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/73687556/a580d401-09f5-427b-89f0-7772ef92aa58)

**Comment by zekker6** [comment]
@tripitakav Yes, it is correct. I've also created a [PR](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/4885) to add support of Loki healthcheck endpoint in order to avoid the confusion, so in future version it will not be required to disable healthchecks.

**Comment by valyala** [comment]
@tripitakav , could you build VictoriaLogs from the commit 8dce4eb18998f06b1f2d15e25dad7afd9345e8a0 according to [these docs](https://docs.victoriametrics.com/VictoriaLogs/QuickStart.html#building-from-source-code) and verify whether it stops panic'ing on your workload?

BTW, the `too big number of columns detected in the block` panic may point to improperly configured data ingestion pipeline. This means that you ingest log entries with big number of fields with unique names - more than 10K in this particular case. Usually log entries contain much smaller number of fields with unique names - up to 100. Big number of field names may point to some kind of misconfiguration at log shipper and pre-processor side. For example, they mistakenly add unique suffixes or prefixes to field names:
```
{
  "fieldName_1": "...",
  "fieldName_2": "...",
  ...,
  "fieldName_10000": "..."
}
```
or
```
{
  "1_fieldName": "...",
  "2_fieldName": "...",
  ...,
  "10000_fieldName": "..."
}
```

You can investigate the actual field names ingested into VictoriaLogs by passing `debug` query arg during data ingestion. In this case VictoriaLogs will log all the ingested log entries with all the field names into stdout. See [these docs](https://docs.victoriametrics.com/VictoriaLogs/data-ingestion/#troubleshooting) for more details.

**Comment by valyala** [comment]
The bugfix has been included in [v0.4.0-victorialogs](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v0.4.0-victoralogs). Closing the issue as fixed then.

@tripitakav , feel free re-opening the issue if it still exists on your workload.

**Comment by tripitakav** [comment]
@valyala OK，Thanks！
